### PR TITLE
[windows] Refactor to optimize vswhere queries

### PR DESCRIPTION
**Pull request by franciscojma86**
## Description
visual_studio.dart is not keeping track whether it has queried vswhere and didn't find an installation, of it hasn't performed any queries at all. This may cause unnecessary queries on the former case. This PR assigns a sentinel value of an empty map to the cached installation details to avoid repeating the queries.

## Related Issues
https://github.com/flutter/flutter/issues/40196

## Tests

I added the following tests:

This is a refactor so the behavior should continue as before.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

### CODE REVIEW

File: packages/flutter_tools/lib/src/windows/visual_studio.dart [hunk 1]
```diff
@@ -18,26 +18,27 @@ class VisualStudio {
...
-  bool get hasNecessaryComponents => _usableVisualStudioDetails != null;
+  bool get hasNecessaryComponents => _usableVisualStudioDetails != null && _usableVisualStudioDetails.isNotEmpty;
 
   /// The name of the Visual Studio install.
   ///
   /// For instance: "Visual Studio Community 2017".
   String get displayName => _bestVisualStudioDetails[_displayNameKey];
 
+  /// The internal catalog map inside the installation details.
+  Map<String, dynamic> get _catalog => _bestVisualStudioDetails[_catalogKey];
```
Comment (stuartmorgan) [inline packages/flutter_tools/lib/src/windows/visual_studio.dart]: Optional: you could simplify the call sites using this by having it return an empty map instead of null; then they don't need to do null checks.

File: packages/flutter_tools/lib/src/windows/visual_studio.dart [hunk 2]
```diff
@@ -72,7 +73,7 @@ class VisualStudio {
   /// the components necessary to build.
   String get vcvarsPath {
     final Map<String, dynamic> details = _usableVisualStudioDetails;
-    if (details == null) {
+    if (details == null || details.isEmpty) {
```
Comment (stuartmorgan) [inline packages/flutter_tools/lib/src/windows/visual_studio.dart]: As above, you only need the second part.

File: packages/flutter_tools/lib/src/windows/visual_studio.dart [hunk 3]
```diff
@@ -201,7 +202,9 @@ class VisualStudio {
   }
 
   /// Returns the details dictionary for the latest version of Visual Studio
-  /// that has all required components.
+  /// that has all required components. If no installation is found, the cached
```
Comment (stuartmorgan) [inline packages/flutter_tools/lib/src/windows/visual_studio.dart]: Let's add ", or {} if there is no such installation" to the end of the first sentence so the return contract is explicit up front.

The second part should be a new paragraph [per Dart style](https://dart.dev/guides/language/effective-dart/documentation#do-separate-the-first-sentence-of-a-doc-comment-into-its-own-paragraph).

Same for the other two *Details getters.

File: packages/flutter_tools/lib/src/windows/visual_studio.dart [hunk 4]
```diff
@@ -210,30 +213,37 @@ class VisualStudio {
     _cachedUsableVisualStudioDetails ??= _visualStudioDetails(
         requiredComponents: _requiredComponents().keys,
         additionalArguments: <String>[_prereleaseKey]);
-    if (_cachedUsableVisualStudioDetails != null) {
-      if (installationHasIssues(_cachedUsableVisualStudioDetails)) {
+    // If an installation with required components if found, but it has issues (e.g. is not complete)
+    // or none is found, assign a sentinel empty value to avoid querying vswhere again.
```
Comment (stuartmorgan) [inline packages/flutter_tools/lib/src/windows/visual_studio.dart]: This is essentially duplicating the note a few lines above; remove one or the other copy.

File: packages/flutter_tools/lib/src/windows/visual_studio.dart [hunk 5]
```diff
@@ -210,30 +213,37 @@ class VisualStudio {
...
+    if (_cachedUsableVisualStudioDetails != null && _cachedUsableVisualStudioDetails.isNotEmpty) {
+      if (installationHasIssues(_cachedUsableVisualStudioDetails)) {  
         _cachedAnyVisualStudioDetails = _cachedUsableVisualStudioDetails;
-        return null;
+        _cachedUsableVisualStudioDetails = <String, dynamic>{};
```
Comment (stuartmorgan) [inline packages/flutter_tools/lib/src/windows/visual_studio.dart]: Optional: I found this re-assignment construction (where the usable version variable is set to a value that might not actually be usable, then unset if needed) a bit hard to follow; I think a local variable for the lookup would make it clearer:
```
if (_cachedUsableVisualStudioDetails == null) {
  Map<String, dynamic> visualStudioWithComponentsDetails = [...];
  visualStudioWithComponentsDetails ??= [...];
  if (visualStudioWithComponentsDetails != null) {
    if (installationHasIssues(visualStudioWithComponentsDetails) {
      _cachedUsable... = visualStudioWithCompontesDetails;
    } else {
      _cachedAny... = visualStudioWithCompontesDetails;
    }
  }
  _cachedUsableVisualStudioDetails ??= <String, dynamic>{};
}
return _cachedUsableVisualStudioDetails;
```

File: packages/flutter_tools/lib/src/windows/visual_studio.dart [hunk 6]
```diff
@@ -210,30 +213,37 @@ class VisualStudio {
...
+    _cachedAnyVisualStudioDetails ??= <String, dynamic>{};
     return _cachedAnyVisualStudioDetails;
   }
 
   /// Returns the details dictionary of the best available version of Visual
   /// Studio. If there's a version that has all the required components, that
   /// will be returned, otherwise returs the lastest installed version (if any).
   Map<String, dynamic> get _bestVisualStudioDetails {
-    if (_usableVisualStudioDetails != null) {
+    if (_usableVisualStudioDetails != null && _usableVisualStudioDetails.isNotEmpty) {
```
Comment (stuartmorgan) [inline packages/flutter_tools/lib/src/windows/visual_studio.dart]: Simplify.

**Review by stuartmorgan-g — CHANGES_REQUESTED** [review]
Thanks! Sorry for writing this technical debt into the implementation in the first place.

**Review by stuartmorgan-g — APPROVED** [review]
